### DISCUSSION TITLE
Is there a way to add custom header to Server Actions calls in Next.js?

### DISCUSSION THREAD

**Discussion by ptvty** [vercel/next.js#74255] [Help]
### Summary

I'm using server actions in my Next.js app (export function myAction() {} with use server). These server actions are invoked from client components.

I want to include a custom request header (e.g., X-Custom-Header) in every request made from the client to these server actions. The goal is for this header to be accessible in my middleware.ts, where I can process it further.

Is there a standard or recommended way to achieve this without globally overriding fetch? Ideally, I’d like to ensure that all server action calls automatically include this header without requiring modifications to every individual action invocation, but it's still okay if there are ways to achieve this with modifications to the function itself.

Thanks!

**Comment by AnouarMc** [comment]
A request made to a server action will include a next-action header by default
![Screenshot_1](https://github.com/user-attachments/assets/5a5c6c49-d6c3-4f9f-83b9-617c6b63beaa)

which you can check in the middleware

```typescript

const isServerAction = request.headers.get("next-action");

**Reply by ptvty** [reply]
I don't want to **get** this header in the **server**, but to **set** a custom header in the **client**!
Thanks for your input

**Reply by AnouarMc** [reply]
As far as I know, there is no way to set a custom header for a server action in the client.

**Comment by yiminghe** [comment]
only by overriding fetch now:

```ts
let action: string = '';
let originalFetch: typeof fetch;
function myFetch(r: any, options?: RequestInit) {
  const newOptions =
    action && !!(options?.headers as any)?.['Next-Action']
      ? {
        ...options,
        headers: {
          ...options?.headers,
          'x-action': action,
        },
      }
      : options;
  return originalFetch(r, newOptions);
}
export async function runWithActionName<T>(
  s: string,
  fn: () => Promise<T>,
): Promise<T> {
  originalFetch = self.fetch;
  self.fetch = myFetch;
  action = s;
  let ret;
  try {
    ret = await fn();
  } finally {
    action = '';
    self.fetch = originalFetch;
  }
  return ret;
}

```
